<b>Комментарий от ревьюера</b>

Привет! Меня зовут Ярослав Кальманов, я буду делать ревью твоего проекта. Давай будем общаться на «ты».

Мои комментарии будут в ячейках "Комментарий от ревьюера".

<div class="alert alert-danger"> 

<b> Комментарий от ревьюера ❗</b>

Красным цветом обозначены комментарии, на которые тебе нужно обратить внимание в первую очередь, чтобы я принял проект. 
</div>
<div class="alert alert-warning">

<b> Комментарий от ревьюера ⚠️</b>

Желтым цветом отмечено то, что ты можешь сделать немного по-другому. Ты можешь учесть эти комментарии при выполнении следующих проектов.
</div>
<div class="alert alert-success">

<b> Комментарий от ревьюера ✔️</b>

Зелёным цветом обозначены твои удачные решения, которые ты без сомнения можешь применять в будущих проектах.  
</div>

Если ты вносишь изменения в проект по моим рекомендациям, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь свои комментарии заметным цветом или поместишь 

<div class="alert alert-info">
    
в такую ячейку
</div>

Давай при работе с проектом мы будем общаться с тобой в диалоге, поэтому не удаляй и не перемещай мои комментарии, это поможет мне сделать повторную  проверку проекта быстрее.
___


# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
debt_from_children = pd.DataFrame()
debt_from_children['count_children'] = data.groupby('children')['debt'].count()
debt_from_children['sum_children'] = data.groupby('children')['debt'].sum()
debt_from_children['result_children'] = debt_from_children['sum_children'] / debt_from_children['count_children']
debt_from_children.sort_values('result_children', ascending = False)

count_children  sum_children  result_children
children                                               
4                     41             4         0.097561
2                   2052           194         0.094542
1                   4808           444         0.092346
3                    330            27         0.081818
0                  14091          1063         0.075438
5                      9             0         0.000000

**Вывод:** 

**На основании полученных данных можно сделать вывод, что бездетные реже имеют просроченные задолженности по кредиту.**

**С увеличением количества детей увеличивается количество просроченных задолженностей. В то же время, люди с 3-я детьми имеют меньше просроченных задолженностей по кредиту, чем люди, имеющие 1-го, 2-х и 4-х детей. Выводы по группам людей, имеющих детей неоднозначны и требуют дополнительного исследования.** 

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Ты хорошо подошла к ответу на вопрос! Здорово, что рассмотрела не только вероятность задолженности внутри групп, но и их размеры. Я бы сказал, что не стоит делать выводы для клиентов с 3+ детей, потому что в группах мало наблюдений, и выводы могут быть не репрезентативны. 
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
debt_from_family_status = pd.DataFrame()
debt_from_family_status['sum_family_status'] = data.groupby('family_status')['debt'].sum()
debt_from_family_status['count_family_status'] = data.groupby('family_status')['debt'].count()
debt_from_family_status['result_family_status'] = debt_from_family_status['sum_family_status'] / debt_from_family_status['count_family_status']
debt_from_family_status.sort_values('result_family_status', ascending = False)

sum_family_status  count_family_status  \
family_status                                                   
Не женат / не замужем                273                 2796   
гражданский брак                     385                 4134   
женат / замужем                      927                12261   
в разводе                             84                 1189   
вдовец / вдова                        63                  951   

                       result_family_status  
family_status                                
Не женат / не замужем              0.097639  
гражданский брак                   0.093130  
женат / замужем                    0.075606  
в разводе                          0.070648  
вдовец / вдова                     0.066246

**Вывод:** 

**Зависимость между семейным положением и возвратом кредита в срок есть.**

**Люди, не состоящие в браке, имеют наибольшее количество просроченных задолженностей. Так же большое количество просроченных задолженностей и у людей, состоящих в гражданском браке.**

**Люди овдовевшие, а также находящиеся в разводе, имеют меньшее количество просроченных задолженностей по кредиту.**

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Все верно 
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data['total_income_id'] = data['total_income'].apply(categorize_income)
data['total_income_id'].value_counts()
debt_from_total_income = pd.DataFrame()
debt_from_total_income['sum_total_income_id'] = data.groupby('total_income_id')['debt'].sum()
debt_from_total_income['count_total_income_id'] = data.groupby('total_income_id')['debt'].count()
debt_from_total_income['conversion_total_income_id'] = debt_from_total_income['sum_total_income_id'] / debt_from_total_income['count_total_income_id']
debt_from_total_income.sort_values('conversion_total_income_id', ascending = False)
                                                               

sum_total_income_id  count_total_income_id  \
total_income_id                                               
E                                  2                     22   
C                               1353                  15921   
A                                  2                     25   
B                                354                   5014   
D                                 21                    349   

                 conversion_total_income_id  
total_income_id                              
E                                  0.090909  
C                                  0.084982  
A                                  0.080000  
B                                  0.070602  
D                                  0.060172

**Вывод:** 

**По уровню дохода клиенты были разделены по категориям:**

- `0 - 30000 - "E"`,
- `30001 - 50000 - "D"`,
- `50001 - 200000 - "C"`,
- `200001 - 1000000 - "B",
- `1000001 и выше - "A"`

**Клиенты категории "Е", относящиеся к категории с самыми низкими доходами, имеют наибольшее количество просроченных задолженностей по кредиту.**

**Клиенты категории "С" и "А" также имеют большое количество просроченных задолженностей. Клиенты категории "D" и "В" реже просрачивают оплату по кредиту.**

**Данные неоднозначные, возможно нужна большая выборка для получения более точного результата.**

<div class="alert alert-warning">
    
<b> Комментарий от ревьюера ⚠️</b>
    
Было бы здорово добавить описание групп в таблицу. Буквенные обозначения удобно использовать при работе с данными, но они не очень информативны. И думаю, не стоит учитывать результаты для групп A, D, E из-за маленького количества клиентов.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
debt_from_purpose_category = pd.DataFrame()
debt_from_purpose_category['sum_purpose_category'] = data.groupby('purpose_category')['debt'].sum()
debt_from_purpose_category['count_purpose_category'] = data.groupby('purpose_category')['debt'].count()
debt_from_purpose_category['result_purpose_category'] = debt_from_purpose_category['sum_purpose_category'] / debt_from_purpose_category['count_purpose_category']
debt_from_purpose_category.sort_values('result_purpose_category', ascending = False)

sum_purpose_category  count_purpose_category  \
purpose_category                                                         
операции с автомобилем                     400                    4279   
получение образования                      369                    3988   
проведение свадьбы                         183                    2313   
операции с недвижимостью                   780                   10751   

                          result_purpose_category  
purpose_category                                   
операции с автомобилем                   0.093480  
получение образования                    0.092528  
проведение свадьбы                       0.079118  
операции с недвижимостью                 0.072551

**Вывод:** 

**Клиенты были разделены на 4 категории в зависимости от целей кредита. 
Меньше всего имеют просроченную задолженность по кредиту клиенты с целью кредита `'операции с недвижимостью'`. Близкие к ним значения и у клиентов с целью кредита `'проведение свадьбы'`.**

**Наибольшее количество просроченных задолженностей по кредиту у клиентов с целью кредита `'операции с автомобилем'` и `'получение образования'`.**


<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Ты хорошо справилась с анализом данных
</div>


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

*В исходных данных были выявлены 2174 пропуска значений в столбцах `'days_employed'` - общий трудовой стаж в днях и `'total_income'` - ежемесячный доход.
Возможной причиной может являться как человеческий фактор, так и технический сбой.
На мой взгляд, ряд клиентов сознательно не заполнили эти данные, так как ежемесячный доход является коммерческой тайной и не все клиенты стремятся его афишировать, а общий трудовой стаж в днях могли не счесть необходимым посчитать. Пропусков в обоих столбцах выявлено одинаковое количество - 2174, можно предположить, что одни и те же клиенты не заполнили оба столбца.*

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Ты права, все это может привести к появлению пропусков.
    
Кстати, у тебя текст слился в один абзац, посмотри [тут](https://paulradzkov.com/2014/markdown_cheatsheet/) про синтаксис Markdown
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

*В выборке могут быть значения, которые значительно отличаются от основной части значений в меньшую или большую сторону. Они могут значительно смещать средний показатель вверх или вниз и влиять на корректность результатов исследования. Заполнение пропусков медианными значениями позволяет получить более точные результаты исследования.*

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Действительно, медиана не зависит от выбросов
</div>


### Шаг 4: общий вывод.

**На вероятность платежей про кредиту в срок влияет семейное положение - люди, состоящие в браке имеют меньше просроченных задолженностей, чем люди в браке не состоящие или состоящие в гражданском браке. Люди овдовевшие и находящие в разводе имеют меньше просроченных заделженностей, чем люди, находящиеся в браке.**

**Количество детей влияет на количество просроченных задолженностей - чем больше детей, тем больше просроченных задолженностей. Бездетные чаще платят в срок. Влияние количества детей на количество просроченных задолженностей, возможно, необходимо изучить на большей выборке и понять, почему люди с 3-мя детьми чаще платят в срок, чем люди с 1-м, 2-мя и 4-мя детьми.**

**Зависимость между уровнем дохода и возвратом кредита в срок, возможно необходимо изучить на большей выборке или увеличив количество исследуемых категорий, например с диапазоном 50000. В нашем случае Клиенты категории "Е", относящиеся к категории с самыми низкими доходами, имеют наибольшее количество просроченных задолженностей по кредиту.Клиенты категории "С" c доходом `50001-200000`и "А" с доходом `1000001 и выше`также имеют большое количество просроченных задолженностей, но меньше, чем клиенты категории "Е". Клиенты категории "D" с доходом `30001-50000`и "В" с доходом `200001-1000000`реже просрачивают оплату по кредиту, чем все остальные категории клиентов.**

**В зависимости от целей кредита меньше всего имеют просроченную задолженность по кредиту клиенты с целью кредита 'операции с недвижимостью'. Близкие к ним значения и у клиентов с целью кредита 'проведение свадьбы'.
Наибольшее количество просроченных задолженностей по кредиту у клиентов с целью кредита 'операции с автомобилем' и 'получение образования'.**


<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Хороший вывод
</div>

<div class="alert alert-success">
    
<b> Комментарий от ревьюера ✔️</b>
    
Спасибо за твою работу! Ты хорошо справилась с анализом данных, у тебя получились логичные комментарии и выводы.

Критических замечаний нет, но я верну работу на случай, если у тебя появились вопросы к моим комментариям. Если вопросов нет, можешь прислать работу без изменений, и я ее приму.
</div>